### PR TITLE
Refactor/ubo step ts migration

### PR DESCRIPTION
**Pull request by barttom**
### Fixed Issues

$
PROPOSAL:

### Tests

- [ ] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [ ] I linked the correct issue in the `### Fixed Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for the expected offline behavior in the `Offline steps` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [ ] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx [hunk 1]
```diff
@@ -27,102 +22,82 @@ import LegalNameUBO from './substeps/BeneficialOwnerDetailsFormSubsteps/LegalNam
...
+type BeneficialOwnerInfoOnyxProps = {
     /** Reimbursement account from ONYX */
-    reimbursementAccount: reimbursementAccountPropTypes,
+    reimbursementAccount: OnyxEntry<ReimbursementAccount>;
 
     /** The draft values of the bank account being setup */
-    reimbursementAccountDraft: reimbursementAccountDraftPropTypes,
-
+    reimbursementAccountDraft: OnyxEntry<ReimbursementAccountDraft>;
+};
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx:31]:
```suggestion
};

```

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx [hunk 2]
```diff
@@ -27,102 +22,82 @@ import LegalNameUBO from './substeps/BeneficialOwnerDetailsFormSubsteps/LegalNam
...
-    reimbursementAccount: ReimbursementAccountProps.reimbursementAccountDefaultProps,
-    reimbursementAccountDraft: {},
-};
+    setIsBeneficialOwnerInfoSet: (newState: boolean) => void;
+} & BeneficialOwnerInfoOnyxProps;
 
-const BODY_CONTENT = [LegalNameUBO, DateOfBirthUBO, SocialSecurityNumberUBO, AddressUBO, ConfirmationUBO];
-const BENEFICIAL_OWNER_INFO_STEP_KEYS = CONST.BANK_ACCOUNT.BENEFICIAL_OWNER_INFO_STEP.INPUT_KEY;
-const BENEFICIAL_OWNER_KEYS_KEY = CONST.BANK_ACCOUNT.BENEFICIAL_OWNER_INFO_STEP.BENEFICIAL_OWNER_DATA.BENEFICIAL_OWNER_KEYS;
+// @ts-expect-error TODO: fix later
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx:43]: is that the time 😄 ?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx [hunk 3]
```diff
@@ -27,102 +22,82 @@ import LegalNameUBO from './substeps/BeneficialOwnerDetailsFormSubsteps/LegalNam
...
-                      firstName: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_firstName`),
-                      lastName: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_lastName`),
-                      dob: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_dob`),
-                      ssnLast4: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_ssnLast4`),
-                      street: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_street`),
-                      city: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_city`),
-                      state: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_state`),
-                      zipCode: lodashGet(reimbursementAccountDraft, `beneficialOwner_${ownerKey}_zipCode`),
-                  }));
+        const beneficialOwnerFields = ['firstName', 'lastName', 'dob', 'ssnLast4', 'street', 'city', 'state', 'zipCode'];
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx:70]: minor but maybe you can move it outside of component scope?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerCheckUBO.tsx [hunk 4]
```diff
@@ -1,73 +1,73 @@
...
-    onSelectedValue: PropTypes.func.isRequired,
+    onSelectedValue: (value: boolean) => void;
 };
 
-function BeneficialOwnerCheckUBO({title, onSelectedValue, defaultValue}) {
+function BeneficialOwnerCheckUBO({title, onSelectedValue, defaultValue}: BeneficialOwnerCheckUBOProps) {
     const {translate} = useLocalize();
     const styles = useThemeStyles();
-    const [value, setValue] = useState(defaultValue);
+    const [value, setValue] = useState(defaultValue.toString());
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerCheckUBO.tsx:24]: why we are changing this to string?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.tsx [hunk 5]
```diff
@@ -1,40 +1,33 @@
...
 const BENEFICIAL_OWNER_INFO_KEY = CONST.BANK_ACCOUNT.BENEFICIAL_OWNER_INFO_STEP.BENEFICIAL_OWNER_DATA;
 const BENEFICIAL_OWNER_PREFIX = CONST.BANK_ACCOUNT.BENEFICIAL_OWNER_INFO_STEP.BENEFICIAL_OWNER_DATA.PREFIX;
 
-function AddressUBO({reimbursementAccountDraft, onNext, isEditing, beneficialOwnerBeingModifiedID}) {
+type AddressUBOOnyxProps = {
+    /** The draft values of the bank account being setup */
+    reimbursementAccountDraft: OnyxEntry<ReimbursementAccountDraft>;
+};
+type AddressUBOProps = SubStepProps & AddressUBOOnyxProps & {beneficialOwnerBeingModifiedID: string};
+type FormValues = BeneficialOwnerDraftData;
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.tsx:24]: is that necessary?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/DateOfBirthUBO.tsx [hunk 6]
```diff
@@ -67,9 +61,11 @@ function DateOfBirthUBO({reimbursementAccountDraft, onNext, isEditing, beneficia
         >
             <Text style={[styles.textHeadline, styles.mb3]}>{translate('beneficialOwnerInfoStep.enterTheDateOfBirthOfTheOwner')}</Text>
             <InputWrapper
+                // @ts-expect-error TODO: Remove this once DatePicker (https://github.com/Expensify/App/issues/25140) is migrated to TypeScript
                 InputComponent={DatePicker}
                 formID={ONYXKEYS.REIMBURSEMENT_ACCOUNT}
                 inputID={dobInputID}
+                z
```
Comment (kubabutkiewicz) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/DateOfBirthUBO.tsx:68]: is this typo?

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.tsx (hunk 5 above)
Comment (barttom) [reply src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/BeneficialOwnerDetailsFormSubsteps/AddressUBO.tsx:24]: I want to keep consistency with other components and show clearly what ty I us for form values

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx (hunk 3 above)
Comment (barttom) [reply src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx:70]: It's strict connected with this function, so I'm not sure what benefits we have move it out of the component. This is simple array, so shouldn't consume a lot of memory ;)

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx [hunk 7]
```diff
@@ -140,7 +122,11 @@ function BeneficialOwnerInfo({reimbursementAccount, reimbursementAccountDraft, o
         prevScreen,
         moveTo,
         resetScreenIndex,
-    } = useSubStep({bodyContent: BODY_CONTENT, startFrom, onFinished: handleBeneficialOwnerDetailsFormSubmit});
+    } = useSubStep<{beneficialOwnerBeingModifiedID: string; setBeneficialOwnerBeingModifiedID?: (id: string) => void}>({
```
Comment (Swor71) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/BeneficialOwnerInfo.tsx:125]: you are using this type in two places, it'd be more readable to extract it to its own type imo

File: src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/CompanyOwnersListUBO.tsx [hunk 8]
```diff
@@ -12,63 +11,83 @@ import Text from '@components/Text';
...
-import * as ReimbursementAccountProps from '@pages/ReimbursementAccount/reimbursementAccountPropTypes';
+import BankAccount from '@libs/models/BankAccount';
 import getSubstepValues from '@pages/ReimbursementAccount/utils/getSubstepValues';
 import getValuesForBeneficialOwner from '@pages/ReimbursementAccount/utils/getValuesForBeneficialOwner';
 import CONST from '@src/CONST';
 import ONYXKEYS from '@src/ONYXKEYS';
+import type {ReimbursementAccount, ReimbursementAccountDraft} from '@src/types/onyx';
 
-const propTypes = {
+const reimbursementAccountDefault = {
```
Comment (Swor71) [inline src/pages/ReimbursementAccount/BeneficialOwnerInfo/substeps/CompanyOwnersListUBO.tsx:21]:
```suggestion
const reimbursementAccountDefault: ReimbursementAccount = {
```

**Comment by mountiny** [comment]
@amyevans feature branch ts migration no need for review at this point

**Pull request merged**